quickwit-cli/src/indexing/document_indexer.rs

```diff
@@ -0,0 +1,130 @@
+
+    let mut split_merge_tasks = vec![];
+
+    let mut current_split = Split::create(&args, schema.clone()).await?;
+    while let Some(raw_doc) = document_receiver.recv().await {
+        let doc_size = raw_doc.as_bytes().len();
+        let parse_result = parse_document(raw_doc);
+        let doc = match parse_result {
+            Ok(doc) => {
+                statistic_sender
+                    .send(StatisticEvent::NewDocument {
+                        size_in_bytes: doc_size,
+                        error: false,
+                    })
+                    .await?;
+                current_split.metadata.num_records += 1;
+                current_split.metadata.size_in_bytes += doc_size;
+                doc
+            }
+            Err(_) => {
+                statistic_sender
+                    .send(StatisticEvent::NewDocument {
+                        size_in_bytes: doc_size,
+                        error: true,
+                    })
+                    .await?;
+                current_split.metadata.num_parsing_errors += 1;
+                continue;
+            }
+        };
+
+        current_split.add_document(doc)?;
+        if current_split.has_enough_docs() {
+            let split = std::mem::replace(
+                &mut current_split,
+                Split::create(&args, schema.clone()).await?,
+            );
+
+            let finalise_split_task = finalize_split(split, statistic_sender.clone()).await?;
+            split_merge_tasks.push(finalise_split_task);
+        }
+    }
+
+    // Finalise last split if it has docs
+    if current_split.metadata.num_records > 0 {
+        let split = std::mem::replace(
+            &mut current_split,
+            Split::create(&args, schema.clone()).await?,
+        );
+
+        let finalise_split_task = finalize_split(split, statistic_sender.clone()).await?;
+        split_merge_tasks.push(finalise_split_task);
+    }
+
+    // for all split finilisation tasks
+    let merge_results = futures::future::try_join_all(split_merge_tasks).await?;
+    //TODO let (splits, errors) = merge_results.iter().partition(|item| item.is_ok());
+
+    let splits = merge_results
+        .into_iter()
+        .filter(|result| result.is_ok())
+        .map(|result| result.unwrap())
+        .collect::<Vec<_>>();
+
+    Ok(splits)
+}
+
+fn parse_document(_raw_doc: String) -> anyhow::Result<Document> {
+    //TODO convert raw json to doc
+    Ok(Document::default())
+}
+
+async fn finalize_split(
+    mut split: Split,
+    statistic_sender: Sender<StatisticEvent>,
+) -> anyhow::Result<JoinHandle<anyhow::Result<Split>>> {
+    statistic_sender
+        .send(StatisticEvent::SplitCreated {
+            id: split.id.to_string(),
+            num_docs: split.metadata.num_records,
+            size_in_bytes: split.metadata.size_in_bytes,
+            num_failed_to_parse_docs: split.metadata.num_parsing_errors,
+        })
+        .await?;
+
+    Ok(task::spawn(async move {
```

Comment:
commit and merge are CPU intensive operation. They should not happen in a regular tokio task. 
Checkout `task::spawn_block` 
https://docs.rs/tokio/1.6.0/tokio/index.html#cpu-bound-tasks-and-blocking-code

(You can ignore the story about rayon. We will reduce the maximum number of threads in the blocking thread pool of tokio)

Reply:
The reason I did not use the spawn_blocking like we did in the barrel repo is that `tantivy::merge_segments`  is not behind a future interface. so we already know this is going to starve the runtime. 
Though I agree this should be the same for `index_writter.commit()` since it is not future-based API. 
But tantivy users should be able to use `index_writer.merge().await` without worrying about this. because this merge interface is already behind a future interface. 
on another note, I thought commit does not do much work. I probably need to dig deeper into that path.

At this stage, it's the commit portion that should go into `tokio::spwn_blocking` i think. and `async index_writer.merge()` should probably make use of `spawn_blocking` internally

Reply:
index_writer.merge() runs the computation on a different thread. The future it returns is a oneshot...

So yes you are right, having it withing a spawned task is ok.

Commit is not cheap at all... It needs to be behind a spawn_blocking.

Reply:
Just saying that the reason I had doubt was that the merge was already returning a future and the commit was not. so this kind of made me doubt since I thought commit was cheap and therefore was left as it is. I have the commit now wrapped in `spawn_blocking`

Reply:
@evanxg852000 if you have a doubt, please investigate the code or ask. 
Committing can have dire consequences, and you cannot rely on code review to catch all of these.

If you do not want to interrupt your flow with that kind of investgation, just drop a TODO that your can revisit after your PR took form.